src/constraint_aware_append.c

```diff
@@ -0,0 +1,384 @@
+}
+
+
+static CustomExecMethods constraint_aware_append_state_methods = {
+	.BeginCustomScan = ca_append_begin,
+	.ExecCustomScan = ca_append_exec,
+	.EndCustomScan = ca_append_end,
+	.ReScanCustomScan = ca_append_rescan,
+	.ExplainCustomScan = ca_append_explain,
+};
+
+static Node *
+constraint_aware_append_state_create(CustomScan *cscan)
+{
+	ConstraintAwareAppendState *state;
+	Append	   *append = linitial(cscan->custom_plans);
+
+	state = (ConstraintAwareAppendState *) newNode(sizeof(ConstraintAwareAppendState),
+												   T_CustomScanState);
+	state->csstate.methods = &constraint_aware_append_state_methods;
+	state->subplan = &append->plan;
+
+	return (Node *) state;
+}
+
+static CustomScanMethods constraint_aware_append_plan_methods = {
+	.CustomName = "ConstraintAwareAppend",
+	.CreateCustomScanState = constraint_aware_append_state_create,
+};
+
+static Plan *
+constraint_aware_append_plan_create(PlannerInfo *root,
+									RelOptInfo *rel,
+									struct CustomPath *path,
+									List *tlist,
+									List *clauses,
+									List *custom_plans)
+{
+	CustomScan *cscan = makeNode(CustomScan);
+	Plan	   *subplan = linitial(custom_plans);
+
+	cscan->scan.scanrelid = 0;	/* Not a real relation we are scanning */
+	cscan->scan.plan.targetlist = tlist;		/* Target list we expect as
+												 * output */
+	cscan->custom_plans = custom_plans;
+	cscan->custom_private = list_make3(list_make1_int(rel->relid),
+									   list_copy(root->append_rel_list),
+									   list_copy(clauses));
+	cscan->custom_scan_tlist = subplan->targetlist;		/* Target list of tuples
+														 * we expect as input */
+	cscan->methods = &constraint_aware_append_plan_methods;
+
+	return &cscan->scan.plan;
+}
+
+static CustomPathMethods constraint_aware_append_path_methods = {
+	.CustomName = "ConstraintAwareAppend",
+	.PlanCustomPath = constraint_aware_append_plan_create,
+};
+
+static inline List *
+remove_parent_subpath(PlannerInfo *root, List *subpaths, Oid parent_relid)
+{
+	Path	   *childpath;
+	Oid			relid;
+
+	childpath = linitial(subpaths);
+	relid = root->simple_rte_array[childpath->parent->relid]->relid;
+
+	if (relid == parent_relid)
+		subpaths = list_delete_first(subpaths);
+
+	return subpaths;
+}
+
+Path *
+constraint_aware_append_path_create(PlannerInfo *root, Hypertable *ht, Path *subpath)
+{
+	ConstraintAwareAppendPath *path;
+	AppendRelInfo *appinfo;
+	Oid			relid;
+
+	path = (ConstraintAwareAppendPath *) newNode(sizeof(ConstraintAwareAppendPath), T_CustomPath);
+	path->cpath.path.pathtype = T_CustomScan;
+	path->cpath.path.rows = subpath->rows;
+	path->cpath.path.startup_cost = subpath->startup_cost;
+	path->cpath.path.total_cost = subpath->total_cost;
+	path->cpath.path.parent = subpath->parent;
+	path->cpath.path.pathkeys = subpath->pathkeys;
+	path->cpath.path.param_info = subpath->param_info;
+	path->cpath.path.pathtarget = subpath->pathtarget;
+	path->cpath.flags = 0;		/* Indicate we support backward scan
```

Comment:
I think it's worth it just for our own knowledge (also for how to test such things) since we are gonna be writing more than one custom scan node. I suspect a backward index scan can trigger this.